### ISSUE TITLE
Compile mexopencv 2.4.x for opencv 2.4.8 and MATLAB R2016a

### ISSUE THREAD

**Issue by phui** [kyamagu/mexopencv#339]
I am trying to compile the packaged code from the [Paper-Doll](http://vision.is.tohoku.ac.jp/~kyamagu/research/paperdoll/) project. Although I am not a seasoned MATLAB user, I do realize there is a big change in 2014 release of Matlab.

After fixing declaration here and there, dropping in new version of bdb, I am left with an issue with compiling mexopencv against opencv 2.4.8.OS is Ubuntu 14.04LTS.

```
Building with 'g++'.
/home/phui/Research/fashion/PaperDoll/paperdoll-v1.0/lib/mexopencv/src/+cv/calcGlobalOrientation.cpp: In function ‘void mexFunction(int, mxArray**, int, const mxArray**)’:
/home/phui/Research/fashion/PaperDoll/paperdoll-v1.0/lib/mexopencv/src/+cv/calcGlobalOrientation.cpp:34:77: error: ‘calcGlobalOrientation’ was not declared in this scope

make: *** [+cv/calcGlobalOrientation.mexa64] Error 255
```

Checking version of opencv in my machine gives:

```
ii  libopencv-calib3d-dev:amd64                           2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-calib3d
ii  libopencv-calib3d2.4:amd64                            2.4.8+dfsg1-2ubuntu1                                amd64        computer vision Camera Calibration library
ii  libopencv-contrib-dev:amd64                           2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-contrib
ii  libopencv-contrib2.4:amd64                            2.4.8+dfsg1-2ubuntu1                                amd64        computer vision contrib library
ii  libopencv-core-dev:amd64                              2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-core
ii  libopencv-core2.4:amd64                               2.4.8+dfsg1-2ubuntu1                                amd64        computer vision core library
ii  libopencv-features2d-dev:amd64                        2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-features2d
ii  libopencv-features2d2.4:amd64                         2.4.8+dfsg1-2ubuntu1                                amd64        computer vision Feature Detection and Descriptor Extraction library
ii  libopencv-flann-dev:amd64                             2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-flann
ii  libopencv-flann2.4:amd64                              2.4.8+dfsg1-2ubuntu1                                amd64        computer vision Clustering and Search in Multi-Dimensional spaces library
rc  libopencv-gpu2.4:amd64                                2.4.8+dfsg1-2ubuntu1                                amd64        computer vision GPU library
ii  libopencv-highgui-dev:amd64                           2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-highgui
ii  libopencv-highgui2.4:amd64                            2.4.8+dfsg1-2ubuntu1                                amd64        computer vision High-level GUI and Media I/O library
ii  libopencv-imgproc-dev:amd64                           2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-imgproc
ii  libopencv-imgproc2.4:amd64                            2.4.8+dfsg1-2ubuntu1                                amd64        computer vision Image Processing library
ii  libopencv-legacy-dev:amd64                            2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-legacy
ii  libopencv-legacy2.4:amd64                             2.4.8+dfsg1-2ubuntu1                                amd64        computer vision legacy library
ii  libopencv-ml-dev:amd64                                2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-ml
ii  libopencv-ml2.4:amd64                                 2.4.8+dfsg1-2ubuntu1                                amd64        computer vision Machine Learning library
ii  libopencv-objdetect-dev:amd64                         2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-objdetect
ii  libopencv-objdetect2.4:amd64                          2.4.8+dfsg1-2ubuntu1                                amd64        computer vision Object Detection library
ii  libopencv-ocl-dev:amd64                               2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-ocl2.4
ii  libopencv-ocl2.4:amd64                                2.4.8+dfsg1-2ubuntu1                                amd64        computer vision OpenCL support library
rc  libopencv-photo2.4:amd64                              2.4.8+dfsg1-2ubuntu1                                amd64        computer vision computational photography library
rc  libopencv-stitching2.4:amd64                          2.4.8+dfsg1-2ubuntu1                                amd64        computer vision image stitching library
rc  libopencv-superres2.4:amd64                           2.4.8+dfsg1-2ubuntu1                                amd64        computer vision Super Resolution library
rc  libopencv-ts2.4:amd64                                 2.4.8+dfsg1-2ubuntu1                                amd64        computer vision ts library
ii  libopencv-video-dev:amd64                             2.4.8+dfsg1-2ubuntu1                                amd64        development files for libopencv-video
ii  libopencv-video2.4:amd64                              2.4.8+dfsg1-2ubuntu1                                amd64        computer vision Video analysis library
rc  libopencv-videostab2.4:amd64                          2.4.8+dfsg1-2ubuntu1                                amd64        computer vision video stabilization library
```

I tried following the instruction in https://github.com/kyamagu/mexopencv/issues/148, but MATLAB throws a tons of error for libopencv_*.so like:

```
ERROR: ld.so: object '/usr/lib/x86_64-linux-gnu/libopencv_gpu.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object '/usr/lib/x86_64-linux-gnu/libopencv_photo.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
```

I feel like I have come a long way only to be defeated by linker issue at the end. Any suggestions?

Many many thanks!

**Comment by kyamagu** [comment]
@phui Are you using the github code or the old packaged TGZ file? If you think mexopencv is the issue, just replace the whole mexopencv directory with the latest release.

**Comment by phui** [comment]
@kyamagu Thank you very much for your prompt reply!

I did start with the old packaged TGZ file, and later replaced it with the "opencv 2.4.X" release of mexopencv. The problem persists after the replacement.

Note that I copied the files in +mexopencv/ to +cv/ to get the dropin replacement to work with PaperDoll compilation script in MATLAB.

**Comment by kyamagu** [comment]
@phui I mean are you using the [github version of paperdoll](https://github.com/kyamagu/paperdoll) or the [old TGZ package](http://vision.is.tohoku.ac.jp/~kyamagu/research/paperdoll/)? You can avoid mexopencv build error by removing unnecessary functions in `+cv`, such as `calcGlobalOrientation`. I remember only a few functions are used in paperdoll, like `warpAffine`.

**Comment by phui** [comment]
@kyamagu Ah, I am using the old TGZ package from the project website. Let me clone the github repo and repeat the whole process tonight.

**Comment by amroamroamro** [comment]
[Last time I looked][1], paperdoll was using only 3 functions from mexopencv. In case you're working with the TGZ package, you're gonna have to replace the mexopencv subfolder with the contents of the [v2.4 branch][2] if you want to compile it against opencv 2.4.8

[1]: https://github.com/kyamagu/mexopencv/issues/275#issuecomment-251937494
[2]: https://github.com/kyamagu/mexopencv/tree/v2.4

**Comment by phui** [comment]
@kyamagu @amroamroamro 
Thank you both so much for the suggestions! I have successfully compiled the current (04/18/2017) version of Paperdoll from github repo.

I followed your instructions to remove unnecessary functions. After removing most cpp files that are stero-related, motion-related and video-related plus some others, the make command in MATLAB successfully compiled 2.4.11 branch of mexopencv against opencv v2.4.8.

**In other words, I kept deleting cpp files in cv+/ folder that generates compilation errors, until it went through.**

I used an image from Fashionista dataset to test the parsing using pretrained model (the 70GB one). Tears almost came out when I see the visualized result after calling the `show_parsing` function.

Again, thank you very much for the suggestions!
